### PR TITLE
New Pregnancy Calculator Goodie

### PR DESCRIPTION
**Pull request by gjamie**
###### What does your Pull Request do (check all that apply)?

Pregnancy Calculator

---

Instant Answer Page: https://duck.co/ia/view/pregnancy_calculator
[gjamie](http://docs.duckduckhack.com/maintaining/guidelines.html): @qofdatabase

### CODE REVIEW

**Comment by daxtheduck** [comment]
@gjamie Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        structured_answer => {
+
+            # ID - Must be unique and match Instant Answer page
+            # E.g. https://duck.co/ia/view/calculator has `id => 'calculator'``
+            id => 'pregnancy_calculator',
+
+            # Name - Used for Answer Bar Tab
+            # Value should be chosen from existing Instant Answer topics
+            # see http://docs.duckduckhack.com/frontend-reference/display-reference.html#name-string-required
+            name => 'Health',
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: Use `Answer`.

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+use DDG::Goodie;
+use strict;
+use DateTime;
+with 'DDG::GoodieRole::Dates';
+
+zci answer_type => 'pregnancy_calculator';
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching`
+# Caching disabled currently. Result depends on the user timezone and will change at midnight.
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: Remove the generated comments.

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 3]
```diff
@@ -0,0 +1,85 @@
+package DDG::Goodie::PregnancyCalculator;
+
+# ABSTRACT: Calculate a due date and how far through pregnancy a person is. If the date is in the future just work out an expected date of delivery
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: We can probably shorten this a bit.

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    my $title="";
+    my $subtitle="";
+    my $retval="";
+    
+#    If the LMP is in the future we can give a projected due date but not say how far through the pregnancy the person is.
+    my $termstring="Full term (40 weeks) : $duedate";
+    if ( $future == 1 ) {
+        $title =$termstring;
+        $retval="Due date (40 weeks) will be $duedate";
+        }    else {
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: Fix spacing.

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    my $currentdate= DateTime->now(time_zone => _get_timezone());
+    my $future= DateTime->compare($lmp,$currentdate); #Is the LMP in the future
+    my $sofar= $lmp->delta_days($currentdate);
+    my $duedate=date_output_string($lmp->add(days=>280));
+    my $title="";
+    my $subtitle="";
+    my $retval="";
+    
+#    If the LMP is in the future we can give a projected due date but not say how far through the pregnancy the person is.
+    my $termstring="Full term (40 weeks) : $duedate";
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: Add spaces either side of operators.

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+              description => "Based on last menstrual period of $lmp_display and cycle length of 28 days."
+              # image => "http://website.com/image.png"
+            },
+
+            meta => {
+                sourceUrl => 'https://en.wikipedia.org/wiki/Estimated_date_of_confinement',
+                sourceName => 'Wikipedia'
+            },
+            input => $lmp_display,
+            operation => "Last menstrual period",
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: I don't think we use this option any more.

File: t/PregnancyCalculator.t [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+zci answer_type => "pregnancy_calculator";
+zci is_cached   => 0;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::PregnancyCalculator )],
+    # At a minimum, be sure to include tests for all:
+    # - primary_example_queries
+    # - secondary_example_queries
+    'lmp 14 Feb 2016' => test_zci(qr/Pregnancy currently at \d{1,2} \+ \d{1} days. Due date \(40 weeks\) will be 20 Nov 2016/,
```
Comment (GuiltyDolphin) [inline t/PregnancyCalculator.t]: Use `re(qr/.../)`.

File: t/PregnancyCalculator.t [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        structured_answer => {
+            id => 'pregnancy_calculator',
+            name => 'Health',
+            meta => {
+                sourceUrl => 'https://en.wikipedia.org/wiki/Estimated_date_of_confinement',
+                sourceName => 'Wikipedia'
+            },
+            input => "14 Feb 2016",
+            operation => "Last menstrual period",
```
Comment (GuiltyDolphin) [inline t/PregnancyCalculator.t]: Use single quotes when interpolation is not required.

File: t/PregnancyCalculator.t [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+zci answer_type => "pregnancy_calculator";
+zci is_cached   => 0;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::PregnancyCalculator )],
+    # At a minimum, be sure to include tests for all:
+    # - primary_example_queries
+    # - secondary_example_queries
+    'lmp 14 Feb 2016' => test_zci(qr/Pregnancy currently at \d{1,2} \+ \d{1} days. Due date \(40 weeks\) will be 20 Nov 2016/,
+        structured_answer => {
```
Comment (GuiltyDolphin) [inline t/PregnancyCalculator.t]: Move the structure into a separate builder so you only need to specify the query and parts of the result that can change.

File: t/PregnancyCalculator.t [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                title => "Full term (40 weeks) : 22 Oct 2030",
+                subtitle => "",
+                description => "Based on last menstrual period of 15 Jan 2030 and cycle length of 28 days."
+                },
+        }),
+
+# Try to include some examples of queries on which it might
+    # appear that your answer will trigger, but does not.
+    'lmp 29 Feb 2015' => undef, #Not a date
+    'lmp last Thursday' => undef, #Not a date that we can deal with
```
Comment (GuiltyDolphin) [inline t/PregnancyCalculator.t]: Remove this (and the above) test; we don't test the formats the Dates Role can parse.

File: t/PregnancyCalculator.t [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+               group => "text",
+            },
+            data => {
+              title => re(qr/Currently \d{1,3} weeks \+ \d{1} days\./),
+              subtitle => "Full term (40 weeks) : 20 Nov 2016",
+              description => "Based on last menstrual period of 14 Feb 2016 and cycle length of 28 days."
+            },
+              
+        }),
+     'last period 15-1-2030' => test_zci('Due date (40 weeks) will be 22 Oct 2030', #test on a future date. Will need updating in2030
```
Comment (GuiltyDolphin) [inline t/PregnancyCalculator.t]: Use ISO8601 format in the tests (e.g., `2030-01-15`).

File: lib/DDG/Goodie/PregnancyCalculator.pm [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+            # Name - Used for Answer Bar Tab
+            # Value should be chosen from existing Instant Answer topics
+            # see http://docs.duckduckhack.com/frontend-reference/display-reference.html#name-string-required
+            name => 'Health',
+
+            data => {
+              title => "$title",
+              subtitle => "$subtitle",
+              description => "Based on last menstrual period of $lmp_display and cycle length of 28 days."
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: This should be in the subtitle.

File: lib/DDG/Goodie/PregnancyCalculator.pm (hunk 6 above)
Comment (mintsoft) [inline lib/DDG/Goodie/PregnancyCalculator.pm]: Yeah this is gone now; is this still some documentation somewhere @gjamie ?

**Comment by mintsoft** [comment]
@gjamie Thanks for the PR; some feedback has been put against the diff. Can I ask where you got the base IA shape (with the comments etc) from? It appears to abide by an older standard for goodies than we're using recently.

File: t/PregnancyCalculator.t (hunk 10 above)
Comment (mintsoft) [inline t/PregnancyCalculator.t]: Although possibly worth having a test for queries that shouldn't fire that aren't dates; for example `lmp definition`

**Comment by gjamie** [comment]
Thanks for the feedback. As you might have guessed I am more a doctor than a coder and not used Perl too much. Most of the output was cut and paste from an existing Goodie (the calculator I think) which is where the operation parameter came from. There is a lot of legacy code!

**Comment by GuiltyDolphin** [comment]
@gjamie Uhh, yeah, I wouldn't use the old Calculator's code - it's been updated in #1927.

You can use `duckpan new` to generate the required boilerplate (some of which can be removed).

**Comment by mintsoft** [comment]
@gjamie Hiya, I'm just checking in, are you still interested in finishing this off? If so do you need any help?

**Comment by pjhampton** [comment]
Looks like this has gone stale. Closing... :+1:

**Pull request closed without merging**